src/v4l/v4lt.rs

```diff
@@ -0,0 +1,184 @@
+use anyhow::{Context, Result};
+use crossbeam_channel::{Receiver, Sender};
+use neolink_core::{
+    bc_protocol::{StreamOutput, StreamOutputError},
+    bcmedia::model::*,
+    Error as NeolinkError,
+};
+use v4l::buffer::Type;
+use v4l::io::traits::OutputStream;
+use v4l::prelude::*;
+use v4l::video::output::Parameters;
+use v4l::video::Output;
+use v4l::{Format, FourCC};
+
+pub(crate) struct V4lDevice {
+    device: Device,
+    receiver: Receiver<BcMedia>,
+    video_width: Option<u32>,
+    video_height: Option<u32>,
+    video_fps: Option<u8>,
+    video_format: Option<StreamFormat>,
+}
+
+pub(crate) struct V4lOutputs {
+    sender: Sender<BcMedia>,
+}
+
+// The stream from the camera will be using one of these formats
+//
+// This is used as part of `StreamOutput` to give hints about
+// the format of the stream
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+enum StreamFormat {
+    // H264 (AVC) video format
+    H264,
+    // H265 (HEVC) video format
+    H265,
+}
+
+impl V4lOutputs {
+    pub(crate) fn new(sender: Sender<BcMedia>) -> Self {
+        Self { sender }
+    }
+}
+
+impl StreamOutput for V4lOutputs {
+    fn write(&mut self, media: BcMedia) -> StreamOutputError {
+        self.sender
+            .send(media)
+            .map_err(|_| NeolinkError::Other("V4l Device dropped"))
+    }
+}
+
+impl V4lDevice {
+    pub(crate) fn from_device(device_index: u8, receiver: Receiver<BcMedia>) -> Result<Self> {
+        let result = Self {
+            device: Device::new(device_index as usize).expect("Failed to create device"),
+            receiver,
+            video_width: None,
+            video_height: None,
+            video_fps: None,
+            video_format: None,
+        };
+        Ok(result)
+    }
+
+    pub(crate) fn run(&mut self) -> Result<()> {
+        // After we have created the device stream we cannot
+        // edit the height/width etc
+        // So first we pull packets from the camera until we have
+        // enough data to setup the height etc
```

Comment:
Is there a way you can query metadata without parsing the video itself? There might be a camera Xml you can fetch that will tell you what its current settings are. (I've seen something similar but I can't recall if it's _available_ or _active_ settings.)

Reply:
This data is also available in
- `<StreamInfoList>` xml (msg id: 146)
- `<Compression>` xml (msg id: 56)
- Initial login (msg id 1)

But this MediaInfo packet was the easiest as the `StreamInfoList` and `Compression` aren't currently programmed in